src/replay/consensus.zig

```diff
@@ -1234,3 +1316,200 @@ test "checkAndHandleNewRoot - success" {
+    const hash2 = SlotAndHash{ .slot = 2, .hash = Hash.initRandom(random) };
+
+    var fixture = try TestFixture.init(testing.allocator, root);
+    defer fixture.deinit(testing.allocator);
+
+    try fixture.fill_keys(testing.allocator, random, 1);
+    const my_vote_pubkey = fixture.vote_pubkeys.items[0];
+
+    // Create the tree: root -> 1, root -> 2
+    var trees1 = try std.BoundedArray(TreeNode, MAX_TEST_TREE_LEN).init(0);
+    trees1.appendSliceAssumeCapacity(&[_]TreeNode{
+        .{ hash1, root },
+        .{ hash2, root },
+    });
+    try fixture.fill_fork(
+        testing.allocator,
+        .{ .root = root, .data = trees1 },
+        true,
+    );
+
+    var ancestors = try convertAncestorsMap(testing.allocator, &fixture.ancestors);
+    defer ancestors.deinit(testing.allocator);
+
+    const versioned_stakes = try testEpochStakes(
+        testing.allocator,
+        fixture.vote_pubkeys.items,
+        10000,
+        random,
+    );
+    defer versioned_stakes.deinit(testing.allocator);
+
+    var epoch_stakes = std.AutoHashMapUnmanaged(Epoch, VersionedEpochStakes).empty;
+    defer epoch_stakes.deinit(testing.allocator);
+    try epoch_stakes.put(testing.allocator, 0, versioned_stakes);
+    try epoch_stakes.put(testing.allocator, 1, versioned_stakes);
+
+    var epoch_schedule = EpochSchedule.DEFAULT;
+    _ = try computeBankStats(
+        testing.allocator,
+        my_vote_pubkey,
+        ancestors,
+        &fixture.slot_tracker,
+        &epoch_stakes,
+        &epoch_schedule,
+        &fixture.progress,
+        &fixture.fork_choice,
+        &fixture.latest_validator_votes_for_frozen_banks,
+    );
+
+    // Check that stake for slot 1 and slot 2 is equal
+    const bank1 = fixture.slot_tracker.get(1).?;
+    const bank2 = fixture.slot_tracker.get(2).?;
+
+    const stake1 = fixture.fork_choice.stakeForSubtree(
+        &.{ .slot = 1, .hash = bank1.state.hash.readCopy().? },
+    ).?;
+    const stake2 = fixture.fork_choice.stakeForSubtree(
+        &.{ .slot = 2, .hash = bank2.state.hash.readCopy().? },
+    ).?;
+    try testing.expectEqual(stake1, stake2);
+
+    // Select the heaviest bank
+    const heaviest = fixture.fork_choice.heaviestOverallSlot();
+    // Should pick the lower of the two equally weighted banks
+    try testing.expectEqual(@as(u64, 1), heaviest.slot);
+}
+
+const builtin = @import("builtin");
+fn convertAncestorsMap(
+    allocator: std.mem.Allocator,
+    src: *const std.AutoArrayHashMapUnmanaged(Slot, SortedSet(Slot)),
+) !std.AutoHashMapUnmanaged(u64, SortedSet(u64)) {
+    if (!builtin.is_test) {
+        @compileError("convertAncestorsMap should only be called in test mode");
+    }
+    var dst = std.AutoHashMapUnmanaged(u64, SortedSet(u64)).empty;
+    try dst.ensureTotalCapacity(allocator, @intCast(src.count()));
+    var it = src.iterator();
+    while (it.next()) |entry| {
+        try dst.put(allocator, entry.key_ptr.*, entry.value_ptr.*);
+    }
+    return dst;
+}
```

Comment:
What does this function actually do? It appears that the input and output are exactly identical.

Reply:
I see, didn't notice that small difference. Any reason the new methods can't just also use array hash maps? If not, I think changing them to that seems preferable to this.

Opinionated side note:
In general, I think non-array hashmaps are kind of a niche data structure, given the sheer amount of code that ends up iterating over hashmaps in general. I think the main situations that justify non-array hashmaps are those where lookup performance is critical.

Reply:
> In general, I think non-array hashmaps are kind of a niche data structure, given the sheer amount of code that ends up iterating over hashmaps in practice.

This is one of the reason why I thought of keeping it as `std.AutoHashMapUnmanaged` in the new methods given all it's usage are lookups, and my understanding is that `std.AutoHashMapUnmanaged` is preferable for lookups while `std.AutoArrayHashMapUnmanaged` is preferably for iterations. - hence why I hesitated from switching the type, but convert instead in the tests, because if I had switched, it would mean the switch was driven by testing, but not the actual runtime behaviour.

Reply:
I would hesitate to make the decision to use non-array hashmap for performance reasons without some empirical data backing up the decision, but I suppose there's not a strong argument against it in this case  either beyond the awkwardness of this facet of testing.

I'm fine to leave it as is if that's how you prefer it, but I would suggest some refactors:
```suggestion
const builtin = @import("builtin");
fn convertAncestorsMap(
    allocator: std.mem.Allocator,
    src: *const std.AutoArrayHashMapUnmanaged(Slot, SortedSet(Slot)),
) !std.AutoHashMapUnmanaged(Slot, SortedSet(Slot)) {
    if (!builtin.is_test) {
        @compileError("convertAncestorsMap should only be called in test mode");
    }
    var dst = std.AutoHashMapUnmanaged(u64, SortedSet(u64)).empty;
    errdefer dst.deinit(allocator);

    try dst.ensureTotalCapacity(allocator, src.count());
    for (src.keys(), src.values()) |key, value| {
        dst.putAssumeCapacity(key, value);
    }
    return dst;
}
```
Addendum: what's the ownership for those values look like? Is it fine that they're not being cloned?